RFC - Standards for testing (#18980)
microsoft/fluentui: rfcs/testing.md

## Background

This RFC came from the discussion of unit testing styles in components such as Flex and Text and the debate of having a valuable test vs a "correct" one.

## Problem statement

At the moment, everyone has their own testing standards and there is no common approach to problems like components that only apply styles and need to be tested.
There is no majority decision on the solution for this problem, being split between Visual Regression tests (VR tests) and Unit Tests (UT).

By defining a common approach, we can make the testing process less ambiguous, as well as provide a guide for the community to follow.

## Detailed Design or Proposal

To start, this is the proposal for defining the goals of each testing layer.

- **Conformance**

  Ensuring the consistency of our components. They test that props like `as`, `ref`, `className` are supported and that the component is rendered.

  These can be achieved simply through the `isConformant` function generated when creating a converged component.

> - **Snapshot**

Reviewer: This is kind of misleading since it's possible to use snapshots for anything that is serializable really
```suggestion
- **DOM Snapshot**
```

>   Stricly DOM structure consistency enforcing. This will guarantee our users don't have breaking changes when they expect a certain structure.

Reviewer 2: ```suggestion
  Stricly DOM structure consistency enforcing. This will guarantee we do not introduce breaking changes for users who expect a certain structure to be rendered by our components.
```

>   Stricly DOM structure consistency enforcing. This will guarantee we do not introduce breaking changes for users who expect a certain structure to be rendered by our components.

Reviewer 3: When do we want to test the DOM structure?
Are we planning to make the DOM structure a part of the API contract and consider any change to DOM a breaking change? That sounds scary.

Author: Regarding the first question, I'd say every component should ensure the consistency of their DOM structure. Should I add that to the RFC?
I personally see value in enforcing this because if we plan for our component library to be extensible in all the different ways we provide, keeping DOM structure stable helps with people who choose to extend it with that in mind.
What specifically about this sounds scary?

>   They are achieved with Jest's `expect(component).toMatchSnapshot()`.

Reviewer 4: can we instead use `toMatchInlineSnapshot()` ? 

while using snapshots is usually an anti-pattern (which also slows down testing execution), when there is a valid use case to use it ( not sure if DOM structure is a valid use-case but that's for different discussion ), inline snaphosts make the test more obvious - you can immediately see what's being asserted etc instead of reaching out to some hidden files (standard snaphosts )

Author: Great suggestion, I'll update it! Please remind me, this gets automatically updated with `jest -u` and it's not something we need to manually update, right?

Also, I think it might be relevant to discuss the use case here. What other benefits do we get from snapshot testing if not for avoiding regressions on a DOM level?

>   They are achieved with Jest's `expect(component).toMatchInlineSnapshot(...)`. This is preferred over `.toMatchSnapshot` as it is immediately explicit what is expected from the test.

Reviewer 5: I'm not sure how practical this is for more complex components since giant inline snapshots (including expanded styles) would make the test files really hard to read. If we want to commit to inline snapshots, we'd need to move nontrivial snapshot tests to separate files and/or place strict limits on what should be snapshot-tested.

Reviewer: IMO giant snapshots are a testing antipattern, and my experience is that reviewers gloss over any giant snapshots. In those cases you should be asserting instead.

Author: I agree, if we have a giant inline snapshot, then something else is wrong. Should that be included in this RFC?

Reviewer 4: this 👇 -> 100%

> IMO giant snapshots are a testing antipattern, and my experience is that reviewers gloss over any giant snapshots. In those cases you should be asserting instead.

- **Unit**

>   Focus on testing the behaviour or functionality. We should use them to guarantee the outcome, wheter this is a styling, behaviour or functionality change.

Reviewer 5: nit
```suggestion
  Focus on testing the behaviour or functionality. We should use them to guarantee the outcome, whether this is a styling, behaviour or functionality change.
```

  We should try to focus on behaviour, when possible, but mainly on the expected results for the code.

>   These are all the additional unit tests defined in your `*.test.ts[x]` file.

Reviewer 4: from experience unit test should be avoided as much as possible with few exceptions:

- non UI related complex helpers that are used in other places

with that said, ideal scenario (for a design system/component/UI) is to author integration like test (enforced by `testing-library` toolkit)
- this creates less brittle tests that gives us good enough guarantees that the functionality of tested UI logic is correct
- usually ends up with less testing changes needed when one adds/removes/fixes features

Where this should happen is up to us to determine
- on artificial environment  / jest + jsdom
- real browser  / cypress + storybook

Author: I'm all for these changes but then we need to fit in the style enforcing somewhere else. Should we go forward with only testing styles with VR tests then?

Reviewer 4: > should we go forward with only testing styles with VR tests then?

probably yes for most use cases ?  again if there is some non trivial logic (a mixin for example) for styles that should be tested on "unit/integration" level

Reviewer 5: > from experience unit test should be avoided as much as possible

We might need to clarify some definitions here--of our existing component tests in a `ComponentName.test.tsx` file, what would you consider to be unit tests vs. integration tests or some other category? Maybe this is imprecise but I'm used to thinking of most of those tests as unit tests, and I'm guessing that's not what you mean.

Reviewer 4: > We might need to clarify some definitions here
true that @ecraig12345 !

Here as quick example (for component space context - UI ):

**Unit test:**
- back in the day -> using enzyme -> testing internal stuff like state etc

**Integration test:**
- testing against "real" DOM -> testing-library basically enforces this kind of approach
- testing ideally not "every" component - but only integrated scenarios that make sense

as a recap this article is nice overview of approaches https://kentcdodds.com/blog/write-tests

> - **Integration**

Reviewer: Having both unit+integration tests seems to be very strange for me, and often even in applications the lines are blurred. I would say that we should just have a single unit test category

Author: I agree, perhaps two categories is misleading but with this RFC I had in mind to cover all kinds of tests and standardize how we do them.
Do you think some rephrasing would help? What I want to say is exactly what you mentioned, our Unit and Integration tests go hand in hand.

Reviewer: Well especially since we currently **don't** have any such thing as integration tests, we should do away with the whole thing until we know what we want from such a thing.

I would rather say that we should mock as little as possible in our unit tests, since we are a library, our dependencies really should be tested as much as possible.

Author: Ah I understand now. By this I meant integration testing internally our dependencies, i.e. `make-styles` without mocking it, and not integration of everything as a whole somewhere.
Would it make sense to remove altogether or merge it in Unit Tests?

Reviewer: My proposal: remove integration altogether and just add that in our unit tests we aim to mock as a little as possible

>   Share focus with Unit tests, but for cases where you want to test your component/service with other components/services, without mocking those dependencies. Currently we don't differentiate them from Unit tests but a possibility would be to split them into separate files. However, given our tight dependency on other utilities on our components (i.e. `make-styles`) such deliniation might be complicated.

Reviewer 2: nit:
```suggestion
  Share focus with Unit tests, but for cases where you want to test your component/service with other components/services, without mocking those dependencies. Currently we don't differentiate them from Unit tests but a possibility would be to split them into separate files. However, given our tight dependency on other utilities on our components (i.e. `make-styles`) such delineation might be complicated.
```

- **Performance**

>   Focus on the rendering performance and avoiding it's degradation. They are currently covered by the [flamegrill](https://github.com/microsoft/Flamegrill) utility.

Reviewer 2: nit:
```suggestion
  Focus on the rendering performance and avoiding its degradation. They are currently covered by the [flamegrill](https://github.com/microsoft/Flamegrill) utility.
```

> - Bundle size

Reviewer 4: ```suggestion
- **Bundle size**
```

  Prevent regressions in the bundle size. Currently measured using our custom solution under `/packages/bundle-size`.

> - Accessibility

Reviewer 4: ```suggestion
- **Accessibility**
```

>   We do not automate accessibility testing as of the writing of this document.

Reviewer 2: I think there is a solution in v0 that we could try to implement or use as a reference to spring up our solution for converged components.

Author: We definitely should look at having some static accessibility test wherever possible. Perhaps @smhigley can help define what can be statically tested?

Reviewer 4: we should adopt industry standard tooling 
- `axe` https://github.com/dequelabs/axe-core-npm
- lint rules https://github.com/jsx-eslint/eslint-plugin-jsx-a11y

nice recap https://web.dev/accessibility-auditing-react/

Reviewer: I would add that testing almost any keyboard/focus requirements for components can only really be done in the E2E phase inside the browser.

Author: I agree with this, accessibility is part of user interaction testing.
Do we have any types of a11y E2E tests to give as an example her?

Reviewer 5: There's also a very old attempt at integrating axe using `axe-puppeteer` in v8 under `apps/a11y-tests`. It ran on all the component examples (not sure if this is the right approach) and saved snapshots of the current errors for each example (again not sure about this approach), but was disabled due to issues with churn in generated classNames and IDs (should have been fixable, but nobody got around to it because IIRC it was basically an FHL project). Not sure how useful this will be going forward but wanted to raise awareness that it exists.

Reviewer: > I agree with this, accessibility is part of user interaction testing.
Do we have any types of a11y E2E tests to give as an example here?

`Menu.e2e.ts` - contains a lot of tests that includes keyboarding scenarios

>   We do not automate accessibility testing as of the writing of this document. E2E tests are used to ensure the component's accessibility requirements are met.

Reviewer 6: I think it might be good to link #17243 here.

> - Visual Regression

Reviewer 4: ```suggestion
- **Visual Regression**
```

  Prevent regressions in component styling. Covered currently with [Screener](https://screener.io/).

> - E2E

Reviewer 4: ```suggestion
- **E2E**
```
